### PR TITLE
Change `asio` BUILD to be compatible with the BCR version

### PR DESCRIPTION
**Pull request by lalten**
Split off https://github.com/mvukov/rules_ros2/pull/238

Change `asio` BUILD to be compatible with the BCR version:
https://github.com/bazelbuild/bazel-central-registry/blob/b58c330b8386c64d38c3587c4a1c55ef344fce14/modules/asio/1.28.2/patches/add_build_file.patch

Note that foxglove_bridge_base was previously transitively depending on `@boringssl//:ssl` through `asio`.

### CODE REVIEW

File: repositories/asio.BUILD.bazel [hunk 1]
```diff
@@ -5,15 +5,10 @@ load("@rules_cc//cc:defs.bzl", "cc_library")
...
-    srcs = glob(["src/*.cpp"]),
-    hdrs = glob([
-        "include/**/*.hpp",
-        "include/**/*.ipp",
+    srcs = glob([
+        "**/*.hpp",
+        "**/*.ipp",
     ]),
-    defines = [
-        "ASIO_SEPARATE_COMPILATION",
```
Comment (mvukov) [inline repositories/asio.BUILD.bazel:14]: What's improvement here?

Comment (lalten) [reply repositories/asio.BUILD.bazel:14]: It's arguably worse to have asio as header only lib actually. But that's the way it's configured in the Bazel Central Registry. So to make sure things will work the same with Bzlmod and Workspace we need to change it to be compatible as we can't change the BCR one (at least it seems like more effort)

Comment (udaya2899) [reply repositories/asio.BUILD.bazel:14]: FWIW, you can create a new version in BCR the way you want, and if the module maintainer agrees, you can have a separate version like 1.28.1.bcr.1. That's what I did for osqp [here](https://github.com/bazelbuild/bazel-central-registry/tree/main/modules/osqp). I wanted a specific flag to be set differently, and I created a 0.6.3.bcr.1

Just saying, so you have the option to change at source.

Comment (mvukov) [reply repositories/asio.BUILD.bazel:14]: I'd vote we keep the impl from this repo as it's better, and upgrade the bzl one once that's fixed.

Comment (lalten) [reply repositories/asio.BUILD.bazel:14]: Updated the BUILD to match https://github.com/bazelbuild/bazel-central-registry/pull/2675/files#diff-d92e7004420fb61b89e25389540f8e1014cdde782379b35bad27adbd04f9e2c9
which in turn is based on this one.

**Review by mvukov — APPROVED** [review]

**Pull request merged**